File: app/src/main/java/com/example/jason/bloodGlucoseMonitoring/StatisticsActivity.java
package com.example.jason.bloodGlucoseMonitoring;

import android.content.SharedPreferences;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.TextView;

import java.util.Calendar;
import java.util.Locale;

import static com.example.jason.bloodGlucoseMonitoring.DBHelper.KEY_MEASUREMENT;
import static com.example.jason.bloodGlucoseMonitoring.DBHelper.KEY_TIME_IN_SECONDS;
import static com.example.jason.bloodGlucoseMonitoring.DBHelper.TABLE_MEASUREMENTS;
import static com.example.jason.bloodGlucoseMonitoring.PreferencesActivity.KEY_PREFS;
import static com.example.jason.bloodGlucoseMonitoring.PreferencesActivity.KEY_PREFS_DIABETES_1TYPE;
import static com.example.jason.bloodGlucoseMonitoring.PreferencesActivity.KEY_PREFS_UNIT_BLOOD_SUGAR_MMOL;
import static com.example.jason.bloodGlucoseMonitoring.PreferencesActivity.unitBloodSugarMmolDefault;

public class StatisticsActivity extends AppCompatActivity {

    public static final int weekInSec = 7 * 24 * 3600; // 7 days
    public static final int monthWithoutWeekInSec = 23 * 24 * 3600; // 30 - 7 days

    // statistics vars
    long startWeekInSec;
    long startMonthInSec;

    // select query vars
    private static final String KEY_COUNT = "COUNT";
    private static final String KEY_AVG = "AVG";
    private static final String KEY_MIN = "MIN";
    private static final String KEY_MAX = "MAX";


    private boolean prefsUnitBloodSugarMmol;
    String sugarFormat;

    Calendar now;

    // views
    TextView tvCountForAllTime;
    TextView tvCountForWeek;
    TextView tvCountForMonth;

    TextView tvForWeekAvg;
    TextView tvForWeekMin;
    TextView tvForWeekMax;

    TextView tvForMonthAvg;
    TextView tvForMonthMin;
    TextView tvForMonthMax;

    TextView tvForAllTimeAvg;
    TextView tvForAllTimeMin;
    TextView tvForAllTimeMax;

    // SQLite database
    DBHelper dbHelper;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_statistics);

        tvCountForAllTime = (TextView) findViewById(R.id.tvCountForAllTime);
        tvCountForWeek = (TextView) findViewById(R.id.tvCountForWeek);
        tvCountForMonth = (TextView) findViewById(R.id.tvCountForMonth);

        tvForWeekAvg = (TextView) findViewById(R.id.tvForWeekAvg);
        tvForWeekMin = (TextView) findViewById(R.id.tvForWeekMin);
        tvForWeekMax = (TextView) findViewById(R.id.tvForWeekMax);

        tvForMonthAvg = (TextView) findViewById(R.id.tvForMonthAvg);
        tvForMonthMin = (TextView) findViewById(R.id.tvForMonthMin);
        tvForMonthMax = (TextView) findViewById(R.id.tvForMonthMax);

        tvForAllTimeAvg = (TextView) findViewById(R.id.tvForAllTimeAvg);
        tvForAllTimeMin = (TextView) findViewById(R.id.tvForAllTimeMin);
        tvForAllTimeMax = (TextView) findViewById(R.id.tvForAllTimeMax);

        dbHelper = new DBHelper(this);
    }

    @Override
    protected void onResume() {
        super.onResume();

        loadPreferences();
        SharedPreferences sharedPref = getSharedPreferences(KEY_PREFS, MODE_PRIVATE);
        prefsUnitBloodSugarMmol = sharedPref.getBoolean(KEY_PREFS_UNIT_BLOOD_SUGAR_MMOL,
                unitBloodSugarMmolDefault);

        if (prefsUnitBloodSugarMmol) sugarFormat = "%1$.1f";
        else sugarFormat = "%1$.1f";

        now = Calendar.getInstance();

        // set starts of day  week in seconds
        now.set(Calendar.HOUR_OF_DAY, 0);
        now.set(Calendar.MINUTE, 0);
        now.set(Calendar.SECOND, 0);

        // get starts week in seconds
        now.setTimeInMillis(now.getTimeInMillis() - weekInSec * 1000);
        startWeekInSec = now.getTimeInMillis() / 1000;

        // get starts month in seconds
        now.setTimeInMillis(now.getTimeInMillis() - monthWithoutWeekInSec * 1000);
        startMonthInSec = now.getTimeInMillis() / 1000;
        /*
        // FUTURE - current week and month sugars
        now.set(Calendar.DAY_OF_WEEK, 2);
        startWeekInSec = now.getTimeInMillis() / 1000;
        now.set(Calendar.DAY_OF_MONTH, 1);
        startMonthInSec = now.getTimeInMillis() / 1000;*/

        loadStatistics();
    }

    // load preferences values
    public void loadPreferences() {
        // get preferences object
        SharedPreferences sharedPref = getSharedPreferences(KEY_PREFS, MODE_PRIVATE);

        // get saved value
        prefsUnitBloodSugarMmol = sharedPref.getBoolean(KEY_PREFS_DIABETES_1TYPE,
                unitBloodSugarMmolDefault);
    }

    // get String query
    public String getStrQuery(String func) {
        return "SELECT " + func + "(" + KEY_MEASUREMENT + ") FROM " + TABLE_MEASUREMENTS;
    }

    // get String query from a given time
    public String getStrQuery(String func, long startTimeInSec) {
        return "SELECT " + func + "(" + KEY_MEASUREMENT + ") FROM " + TABLE_MEASUREMENTS
                + " WHERE " + KEY_TIME_IN_SECONDS + " > " + startTimeInSec;
    }

    // load statistics
    private void loadStatistics() {
        SQLiteDatabase database = dbHelper.getWritableDatabase();

        // statistics vars ---------------------start
        //  --counts
        int countAllTime;
        int countWeek;
        int countMonth;

        // weeks sugars
        float weekSugarAvg;
        float weekSugarMin;
        float weekSugarMax;

        // months sugars
        float monthSugarAvg;
        float monthSugarMin;
        float monthSugarMax;

        // all times sugars
        float allTimeSugarAvg;//
        float allTimeSugarMin;//
        float allTimeSugarMax;//
        // statistics vars ---------------------end

        // get all records count
        Cursor cursor = database.rawQuery(getStrQuery(KEY_COUNT), null);
        cursor.moveToFirst();
        countAllTime = cursor.getInt(0);

        // set Info to textViews
        tvCountForAllTime.setText(String.valueOf(countAllTime));

        if (countAllTime > 0) {

            // work with count
            // average
            cursor = database.rawQuery(getStrQuery(KEY_COUNT, startWeekInSec), null);
            cursor.moveToFirst();
            countWeek = prefsUnitBloodSugarMmol ? cursor.getInt(0) : cursor.getInt(0) * 18;

            // minimum
            cursor = database.rawQuery(getStrQuery(KEY_COUNT, startMonthInSec), null);
            cursor.moveToFirst();
            countMonth = prefsUnitBloodSugarMmol ? cursor.getInt(0) : cursor.getInt(0) * 18;

            // set Info to textViews
            tvCountForWeek.setText(String.valueOf(countWeek));
            tvCountForMonth.setText(String.valueOf(countMonth));

            // work with week sugars
            if (countWeek > 0) {
                // average
                cursor = database.rawQuery(getStrQuery(KEY_AVG, startWeekInSec), null);
                cursor.moveToFirst();
                weekSugarAvg = prefsUnitBloodSugarMmol ? cursor.getFloat(0) : cursor.getFloat(0) * 18;

                // minimum
                cursor = database.rawQuery(getStrQuery(KEY_MIN, startWeekInSec), null);
                cursor.moveToFirst();
                weekSugarMin = prefsUnitBloodSugarMmol ? cursor.getFloat(0) : cursor.getFloat(0) * 18;

                // maximum
                cursor = database.rawQuery(getStrQuery(KEY_MAX, startWeekInSec), null);
                cursor.moveToFirst();
                weekSugarMax = prefsUnitBloodSugarMmol ? cursor.getFloat(0) : cursor.getFloat(0) * 18;

                // set Info to textViews
                tvForWeekAvg.setText(String.format(Locale.ENGLISH, sugarFormat, weekSugarAvg));
                tvForWeekMin.setText(String.format(Locale.ENGLISH, sugarFormat, weekSugarMin));
                tvForWeekMax.setText(String.format(Locale.ENGLISH, sugarFormat, weekSugarMax));
            }

            // work with month sugars
            if (countWeek > 0) {
                // average
                cursor = database.rawQuery(getStrQuery(KEY_AVG, startMonthInSec), null);
                cursor.moveToFirst();
                monthSugarAvg = prefsUnitBloodSugarMmol ? cursor.getFloat(0) : cursor.getFloat(0) * 18;

                // minimum
                cursor = database.rawQuery(getStrQuery(KEY_MIN, startMonthInSec), null);
                cursor.moveToFirst();
                monthSugarMin = prefsUnitBloodSugarMmol ? cursor.getFloat(0) : cursor.getFloat(0) * 18;

                // maximum
                cursor = database.rawQuery(getStrQuery(KEY_MAX, startMonthInSec), null);
                cursor.moveToFirst();
                monthSugarMax = prefsUnitBloodSugarMmol ? cursor.getFloat(0) : cursor.getFloat(0) * 18;

                // set Info to textViews
                tvForMonthAvg.setText(String.format(Locale.ENGLISH, sugarFormat, monthSugarAvg));
                tvForMonthMin.setText(String.format(Locale.ENGLISH, sugarFormat, monthSugarMin));
                tvForMonthMax.setText(String.format(Locale.ENGLISH, sugarFormat, monthSugarMax));
            }

            // work with all time sugars
            // average
            cursor = database.rawQuery(getStrQuery(KEY_AVG), null);
            cursor.moveToFirst();
            allTimeSugarAvg = prefsUnitBloodSugarMmol ? cursor.getFloat(0) : cursor.getFloat(0) * 18;

            // minimum
            cursor = database.rawQuery(getStrQuery(KEY_MIN), null);
            cursor.moveToFirst();
            allTimeSugarMin = prefsUnitBloodSugarMmol ? cursor.getFloat(0) : cursor.getFloat(0) * 18;

            // maximum
            cursor = database.rawQuery(getStrQuery(KEY_MAX), null);
            cursor.moveToFirst();
            allTimeSugarMax = prefsUnitBloodSugarMmol ? cursor.getFloat(0) : cursor.getFloat(0) * 18;

            // set Info to textViews
            tvForAllTimeAvg.setText(String.format(Locale.ENGLISH, sugarFormat, allTimeSugarAvg));
            tvForAllTimeMin.setText(String.format(Locale.ENGLISH, sugarFormat, allTimeSugarMin));
            tvForAllTimeMax.setText(String.format(Locale.ENGLISH, sugarFormat, allTimeSugarMax));
        }
        cursor.close();
        database.close();
    }
}

File: app/src/main/java/ru/opalevapps/EveryGlic/ui/StatisticsActivity.kt
package ru.opalevapps.EveryGlic.ui

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import ru.opalevapps.EveryGlic.MyUtill
import ru.opalevapps.EveryGlic.R
import ru.opalevapps.EveryGlic.db.DBHelper
import java.text.SimpleDateFormat
import java.util.*

class StatisticsActivity : AppCompatActivity() {
    // statistics vars
    private var startLastWeekInSec: Long = 0
    private var startCurWeekInSec: Long = 0
    private var startLastMonthInSec: Long = 0
    private var startCurMonthInSec: Long = 0

    // variables for preferences
    private var prefsUnitBloodSugarMmol = false
    private var prefsBloodLowSugar = 0f
    private var prefsBloodHighSugar = 0f
    private var prefsBeginningWeek = 0

    // For date and time
    private var sugarFormat: String? = null
    private var now: Calendar? = null

    // views
    private var tvAllTimeCount: TextView? = null
    private var tvAllTimeCountLow: TextView? = null
    private var tvAllTimeCountHigh: TextView? = null
    private var tvAllTimeAvg: TextView? = null
    private var tvAllTimeMin: TextView? = null
    private var tvAllTimeMax: TextView? = null
    private var tvCurWeekCount: TextView? = null
    private var tvCurWeekCountLow: TextView? = null
    private var tvCurWeekCountHigh: TextView? = null
    private var tvCurWeekAvg: TextView? = null
    private var tvCurWeekMin: TextView? = null
    private var tvCurWeekMax: TextView? = null
    private var tvCurMonthCount: TextView? = null
    private var tvCurMonthCountLow: TextView? = null
    private var tvCurMonthCountHigh: TextView? = null
    private var tvCurMonthAvg: TextView? = null
    private var tvCurMonthMin: TextView? = null
    private var tvCurMonthMax: TextView? = null
    private var tvLastWeekCount: TextView? = null
    private var tvLastWeekCountLow: TextView? = null
    private var tvLastWeekCountHigh: TextView? = null
    private var tvLastWeekAvg: TextView? = null
    private var tvLastWeekMin: TextView? = null
    private var tvLastWeekMax: TextView? = null
    private var tvLastMonthCount: TextView? = null
    private var tvLastMonthCountLow: TextView? = null
    private var tvLastMonthCountHigh: TextView? = null
    private var tvLastMonthAvg: TextView? = null
    private var tvLastMonthMin: TextView? = null
    private var tvLastMonthMax: TextView? = null

    // SQLite database
    private var dbHelper: DBHelper? = null

    // display date format
    private var DATE_FORMAT = "dd/MM EEE- "
    private var simpleDateFormat = SimpleDateFormat(DATE_FORMAT)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_statistics)
        initViews()
        dbHelper = DBHelper(this)
    }

    override fun onResume() {
        super.onResume()
        loadPreferences()

        // set display sugar format
        sugarFormat = if (prefsUnitBloodSugarMmol) "%1$.1f" else "%1$.0f"

        /*if (prefsUnitBloodSugarMmol) sugarFormat = "%1$.1f";
        else sugarFormat = "%1$.0f";*/

        // get starts last week in seconds
        resetNowDate()
        now!![Calendar.DAY_OF_YEAR] = now!![Calendar.DAY_OF_YEAR] - 7
        startLastWeekInSec = now!!.timeInMillis / 1000

        // get starts current week in seconds
        resetNowDate()
        now!![Calendar.DAY_OF_WEEK] = prefsBeginningWeek
        if (prefsBeginningWeek < 2) now!![Calendar.DAY_OF_YEAR] = now!![Calendar.DAY_OF_YEAR] - 7
        startCurWeekInSec = now!!.timeInMillis / 1000

        // get starts current month in seconds
        resetNowDate()
        now!![Calendar.DAY_OF_MONTH] = 1
        startCurMonthInSec = now!!.timeInMillis / 1000

        // get starts last month in seconds
        resetNowDate()
        now!![Calendar.DAY_OF_YEAR] = now!![Calendar.DAY_OF_YEAR] - 30
        startLastMonthInSec = now!!.timeInMillis / 1000
        Log.d(
            TAG,
            "StatisticsActivity, onResume: " + simpleDateFormat.format(startCurWeekInSec * 1000) + prefsBeginningWeek
        )
        loadStatistics()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        MyUtill.createInfoItemInActionBar(menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        MyUtill.parseMenuItemInfo(this, item)
        return super.onOptionsItemSelected(item)
    }

    // load preferences values
    fun resetNowDate() {
        now = Calendar.getInstance()
        // set starts of day  week in seconds
        now?.set(Calendar.HOUR_OF_DAY, 0)
        now?.set(Calendar.MINUTE, 0)
        now?.set(Calendar.SECOND, 0)
    }

    // load preferences values
    fun loadPreferences() {
        // get preferences object
        val sharedPref = getSharedPreferences(PreferencesActivity.KEY_PREFS, MODE_PRIVATE)

        // get saved value
        prefsUnitBloodSugarMmol = sharedPref.getBoolean(
            PreferencesActivity.KEY_PREFS_UNIT_BLOOD_SUGAR_MMOL,
            PreferencesActivity.UNIT_BLOOD_SUGAR_MMOL_DEFAULT
        )
        prefsBloodLowSugar = sharedPref.getFloat(
            PreferencesActivity.KEY_PREFS_BLOOD_LOW_SUGAR,
            PreferencesActivity.BLOOD_LOW_SUGAR_DEFAULT
        )
        prefsBloodHighSugar = sharedPref.getFloat(
            PreferencesActivity.KEY_PREFS_BLOOD_HIGH_SUGAR,
            PreferencesActivity.BLOOD_HIGH_SUGAR_DEFAULT
        )
        prefsBeginningWeek = sharedPref.getInt(
            PreferencesActivity.KEY_PREFS_BEGINNING_WEEK,
            PreferencesActivity.BEGINNING_WEEK_DEFAULT
        )
    }

    // get String query
    fun getStrQuery(func: String): String {
        return "SELECT " + func + "(" + DBHelper.KEY_MEASUREMENT + ") FROM " + DBHelper.TABLE_MEASUREMENTS
    }

    // get String query from a given time
    fun getStrQuery(func: String, boundarySugar: Float, lowSugar: Boolean): String {
        val equalitySign: String
        equalitySign = if (lowSugar) " < " else " > "
        return ("SELECT " + func + "(" + DBHelper.KEY_MEASUREMENT + ") FROM " + DBHelper.TABLE_MEASUREMENTS
                + " WHERE " + DBHelper.KEY_MEASUREMENT + equalitySign + boundarySugar)
    }

    // get String query from a given time
    fun getStrQuery(func: String, startTimeInSec: Long): String {
        return ("SELECT " + func + "(" + DBHelper.KEY_MEASUREMENT + ") FROM " + DBHelper.TABLE_MEASUREMENTS
                + " WHERE " + DBHelper.KEY_TIME_IN_SECONDS + " > " + startTimeInSec)
    }

    // get String query from a given time
    fun getStrQuery(
        func: String,
        startTimeInSec: Long,
        boundarySugar: Float,
        lowSugar: Boolean
    ): String {
        val equalitySign: String
        equalitySign = if (lowSugar) " < " else " > "
        return ("SELECT " + func + "(" + DBHelper.KEY_MEASUREMENT + ") FROM " + DBHelper.TABLE_MEASUREMENTS
                + " WHERE " + DBHelper.KEY_TIME_IN_SECONDS + " > " + startTimeInSec
                + " AND " + DBHelper.KEY_MEASUREMENT + equalitySign + boundarySugar)
    }

    // load statistics
    private fun loadStatistics() {
        val database = dbHelper!!.writableDatabase

        // statistics vars ---------------------start
        // all times sugars
        val allTimeCount: Int
        val allTimeCountLow: Int
        val allTimeCountHigh: Int
        var allTimeSugarAvg: Float
        var allTimeSugarMin: Float
        var allTimeSugarMax: Float

        // last weeks sugars
        val curWeekCount: Int
        val curWeekCountLow: Int
        val curWeekCountHigh: Int
        var curWeekSugarAvg: Float
        var curWeekSugarMin: Float
        var curWeekSugarMax: Float

        // last months sugars
        val curMonthCount: Int
        val curMonthCountLow: Int
        val curMonthCountHigh: Int
        var curMonthSugarAvg: Float
        var curMonthSugarMin: Float
        var curMonthSugarMax: Float

        // last weeks sugars
        val lastWeekCount: Int
        val lastWeekCountLow: Int
        val lastWeekCountHigh: Int
        var lastWeekSugarAvg: Float
        var lastWeekSugarMin: Float
        var lastWeekSugarMax: Float

        // last months sugars
        val lastMonthCount: Int
        val lastMonthCountLow: Int
        val lastMonthCountHigh: Int
        var lastMonthSugarAvg: Float
        var lastMonthSugarMin: Float
        var lastMonthSugarMax: Float
        // statistics vars ---------------------end

        // get all records count
        var cursor = database.rawQuery(getStrQuery(KEY_COUNT), null)
        cursor.moveToFirst()
        allTimeCount = cursor.getInt(0)

        // set Info to textViews
        tvAllTimeCount!!.text = allTimeCount.toString()

        // if measurement doesn't exist then don't try get statistics data
        if (allTimeCount > 0) {
            // work with count
            // current week
            cursor = database.rawQuery(getStrQuery(KEY_COUNT, startCurWeekInSec), null)
            cursor.moveToFirst()
            curWeekCount = cursor.getInt(0)
            cursor.close()

            // current month
            cursor = database.rawQuery(getStrQuery(KEY_COUNT, startCurMonthInSec), null)
            cursor.moveToFirst()
            curMonthCount = cursor.getInt(0)
            cursor.close()

            // last week
            cursor = database.rawQuery(getStrQuery(KEY_COUNT, startLastWeekInSec), null)
            cursor.moveToFirst()
            lastWeekCount = cursor.getInt(0)
            cursor.close()

            // last month
            cursor = database.rawQuery(getStrQuery(KEY_COUNT, startLastMonthInSec), null)
            cursor.moveToFirst()
            lastMonthCount = cursor.getInt(0)
            cursor.close()

            // set Info to textViews
            tvCurWeekCount!!.text = curWeekCount.toString()
            tvCurMonthCount!!.text = curMonthCount.toString()
            tvLastWeekCount!!.text = lastWeekCount.toString()
            tvLastMonthCount!!.text = lastMonthCount.toString()

            // work with last week sugars
            if (lastWeekCount > 0) {
                // average
                cursor = database.rawQuery(getStrQuery(KEY_AVG, startLastWeekInSec), null)
                cursor.moveToFirst()
                lastWeekSugarAvg = cursor.getFloat(0)
                cursor.close()

                // minimum
                cursor = database.rawQuery(getStrQuery(KEY_MIN, startLastWeekInSec), null)
                cursor.moveToFirst()
                lastWeekSugarMin = cursor.getFloat(0)
                cursor.close()

                // maximum
                cursor = database.rawQuery(getStrQuery(KEY_MAX, startLastWeekInSec), null)
                cursor.moveToFirst()
                lastWeekSugarMax = cursor.getFloat(0)
                cursor.close()

                // to display measurement in selected in preferences unit of sugar measurement
                if (!prefsUnitBloodSugarMmol) {
                    lastWeekSugarAvg *= 18f
                    lastWeekSugarMin *= 18f
                    lastWeekSugarMax *= 18f
                }

                // set Info to textViews
                tvLastWeekAvg!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, lastWeekSugarAvg)
                tvLastWeekMin!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, lastWeekSugarMin)
                tvLastWeekMax!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, lastWeekSugarMax)
            }

            // work with last month sugars
            if (lastMonthCount > 0) {
                // average
                cursor = database.rawQuery(getStrQuery(KEY_AVG, startLastMonthInSec), null)
                cursor.moveToFirst()
                lastMonthSugarAvg = cursor.getFloat(0)
                cursor.close()

                // minimum
                cursor = database.rawQuery(getStrQuery(KEY_MIN, startLastMonthInSec), null)
                cursor.moveToFirst()
                lastMonthSugarMin = cursor.getFloat(0)
                cursor.close()

                // maximum
                cursor = database.rawQuery(getStrQuery(KEY_MAX, startLastMonthInSec), null)
                cursor.moveToFirst()
                lastMonthSugarMax = cursor.getFloat(0)
                cursor.close()

                // to display measurement in selected in preferences unit of sugar measurement
                if (!prefsUnitBloodSugarMmol) {
                    lastMonthSugarAvg *= 18f
                    lastMonthSugarMin *= 18f
                    lastMonthSugarMax *= 18f
                }

                // set Info to textViews
                tvLastMonthAvg!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, lastMonthSugarAvg)
                tvLastMonthMin!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, lastMonthSugarMin)
                tvLastMonthMax!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, lastMonthSugarMax)
            }

            // current week sugars work
            if (curWeekCount > 0) {
                // average
                cursor = database.rawQuery(getStrQuery(KEY_AVG, startCurMonthInSec), null)
                cursor.moveToFirst()
                curWeekSugarAvg = cursor.getFloat(0)
                cursor.close()

                // minimum
                cursor = database.rawQuery(getStrQuery(KEY_MIN, startCurWeekInSec), null)
                cursor.moveToFirst()
                curWeekSugarMin = cursor.getFloat(0)
                cursor.close()

                // maximum
                cursor = database.rawQuery(getStrQuery(KEY_MAX, startCurWeekInSec), null)
                cursor.moveToFirst()
                curWeekSugarMax = cursor.getFloat(0)
                cursor.close()

                // to display measurement in selected in preferences unit of sugar measurement
                if (!prefsUnitBloodSugarMmol) {
                    curWeekSugarAvg *= 18f
                    curWeekSugarMin *= 18f
                    curWeekSugarMax *= 18f
                }

                // set Info to textViews
                tvCurWeekAvg!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, curWeekSugarAvg)
                tvCurWeekMin!!.text = String.format(Locale.ENGLISH, sugarFormat!!, curWeekSugarMin)
                tvCurWeekMax!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, curWeekSugarMax)
            }

            //  current month sugars work
            if (curMonthCount > 0) {
                // average
                cursor = database.rawQuery(getStrQuery(KEY_AVG, startCurMonthInSec), null)
                cursor.moveToFirst()
                curMonthSugarAvg = cursor.getFloat(0)
                cursor.close()

                // minimum
                cursor = database.rawQuery(getStrQuery(KEY_MIN, startCurMonthInSec), null)
                cursor.moveToFirst()
                curMonthSugarMin = cursor.getFloat(0)
                cursor.close()

                // maximum
                cursor = database.rawQuery(getStrQuery(KEY_MAX, startCurMonthInSec), null)
                cursor.moveToFirst()
                curMonthSugarMax = cursor.getFloat(0)
                cursor.close()

                // to display measurement in selected in preferences unit of sugar measurement
                if (!prefsUnitBloodSugarMmol) {
                    curMonthSugarAvg *= 18f
                    curMonthSugarMin *= 18f
                    curMonthSugarMax *= 18f
                }

                // set Info to textViews
                tvCurMonthAvg!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, curMonthSugarAvg)
                tvCurMonthMin!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, curMonthSugarMin)
                tvCurMonthMax!!.text =
                    String.format(Locale.ENGLISH, sugarFormat!!, curMonthSugarMax)
            }

            // all time sugars work
            // average
            cursor = database.rawQuery(getStrQuery(KEY_AVG), null)
            cursor.moveToFirst()
            allTimeSugarAvg = cursor.getFloat(0)
            cursor.close()

            // minimum
            cursor = database.rawQuery(getStrQuery(KEY_MIN), null)
            cursor.moveToFirst()
            allTimeSugarMin = cursor.getFloat(0)
            cursor.close()

            // maximum
            cursor = database.rawQuery(getStrQuery(KEY_MAX), null)
            cursor.moveToFirst()
            allTimeSugarMax = cursor.getFloat(0)
            cursor.close()

            // to display measurement in selected in preferences unit of sugar measurement
            if (!prefsUnitBloodSugarMmol) {
                allTimeSugarAvg *= 18f
                allTimeSugarMin *= 18f
                allTimeSugarMax *= 18f
            }

            // set Info to textViews
            tvAllTimeAvg!!.text =
                String.format(Locale.ENGLISH, sugarFormat!!, allTimeSugarAvg)
            tvAllTimeMin!!.text = String.format(Locale.ENGLISH, sugarFormat!!, allTimeSugarMin)
            tvAllTimeMax!!.text =
                String.format(Locale.ENGLISH, sugarFormat!!, allTimeSugarMax)


            // count of low and high sugars for all time
            cursor =
                database.rawQuery(getStrQuery(KEY_COUNT, prefsBloodLowSugar, KEY_LOW_SUGAR), null)
            cursor.moveToFirst()
            allTimeCountLow = cursor.getInt(0)
            cursor.close()

            cursor =
                database.rawQuery(getStrQuery(KEY_COUNT, prefsBloodHighSugar, KEY_HIGH_SUGAR), null)
            cursor.moveToFirst()
            allTimeCountHigh = cursor.getInt(0)
            cursor.close()

            // set Info to textViews
            tvAllTimeCountLow!!.text = allTimeCountLow.toString()
            tvAllTimeCountHigh!!.text = allTimeCountHigh.toString()

            //  count of low sugars work
            if (allTimeCountLow > 0) {
                // current month low sugars
                cursor = database.rawQuery(
                    getStrQuery(
                        KEY_COUNT,
                        startCurMonthInSec,
                        prefsBloodLowSugar,
                        KEY_LOW_SUGAR
                    ), null
                )
                cursor.moveToFirst()
                curMonthCountLow = cursor.getInt(0)
                cursor.close()

                // last month low sugars
                cursor = database.rawQuery(
                    getStrQuery(
                        KEY_COUNT,
                        startLastMonthInSec,
                        prefsBloodLowSugar,
                        KEY_LOW_SUGAR
                    ), null
                )
                cursor.moveToFirst()
                lastMonthCountLow = cursor.getInt(0)
                cursor.close()

                // current week low sugars
                if (curMonthCountLow > 0) {
                    cursor = database.rawQuery(
                        getStrQuery(
                            KEY_COUNT,
                            startCurWeekInSec,
                            prefsBloodLowSugar,
                            KEY_LOW_SUGAR
                        ), null
                    )
                    cursor.moveToFirst()
                    curWeekCountLow = cursor.getInt(0)
                } else curWeekCountLow = 0

                // last week low sugars
                if (lastMonthCountLow > 0) {
                    cursor = database.rawQuery(
                        getStrQuery(
                            KEY_COUNT,
                            startLastWeekInSec,
                            prefsBloodLowSugar,
                            KEY_LOW_SUGAR
                        ), null
                    )
                    cursor.moveToFirst()
                    lastWeekCountLow = cursor.getInt(0)
                } else lastWeekCountLow = 0

                // set count of low sugars Info to textViews
                tvCurWeekCountLow!!.text = curWeekCountLow.toString()
                tvLastWeekCountLow!!.text = lastWeekCountLow.toString()
            } else {
                curMonthCountLow = 0
                lastMonthCountLow = 0
            }

            //  count of high sugars work
            if (allTimeCountHigh > 0) {
                // current month high sugars
                cursor = database.rawQuery(
                    getStrQuery(
                        KEY_COUNT,
                        startCurMonthInSec,
                        prefsBloodHighSugar,
                        KEY_HIGH_SUGAR
                    ), null
                )
                cursor.moveToFirst()
                curMonthCountHigh = cursor.getInt(0)
                cursor.close()

                // last month high sugars
                cursor = database.rawQuery(
                    getStrQuery(
                        KEY_COUNT,
                        startLastMonthInSec,
                        prefsBloodHighSugar,
                        KEY_HIGH_SUGAR
                    ), null
                )
                cursor.moveToFirst()
                lastMonthCountHigh = cursor.getInt(0)
                cursor.close()

                // current week high sugars
                if (curMonthCountHigh > 0) {
                    cursor = database.rawQuery(
                        getStrQuery(
                            KEY_COUNT,
                            startCurWeekInSec,
                            prefsBloodHighSugar,
                            KEY_HIGH_SUGAR
                        ), null
                    )
                    cursor.moveToFirst()
                    curWeekCountHigh = cursor.getInt(0)
                    cursor.close()
                } else curWeekCountHigh = 0

                // last week high sugars
                if (lastMonthCountHigh > 0) {
                    cursor = database.rawQuery(
                        getStrQuery(
                            KEY_COUNT,
                            startLastWeekInSec,
                            prefsBloodHighSugar,
                            KEY_HIGH_SUGAR
                        ), null
                    )
                    cursor.moveToFirst()
                    lastWeekCountHigh = cursor.getInt(0)
                    cursor.close()
                } else lastWeekCountHigh = 0

                // set count of high sugars Info to textViews
                tvCurWeekCountHigh!!.text = curWeekCountHigh.toString()
                tvLastWeekCountHigh!!.text = lastWeekCountHigh.toString()
            } else {
                curMonthCountHigh = 0
                lastMonthCountHigh = 0
            }

            // set count of low and high sugars Info to textViews
            tvCurMonthCountLow!!.text = curMonthCountLow.toString()
            tvLastMonthCountLow!!.text = lastMonthCountLow.toString()
            tvCurMonthCountHigh!!.text = curMonthCountHigh.toString()
            tvLastMonthCountHigh!!.text = lastMonthCountHigh.toString()
        }
        cursor.close()
        database.close()
    }

    // initialize views on screen and their listening
    fun initViews() {
        // find views on screen by id
        tvAllTimeCount = findViewById(R.id.tvCountAllTime)
        tvAllTimeCountLow = findViewById(R.id.tvLastAllTimeCountLow)
        tvAllTimeCountHigh = findViewById(R.id.tvLastAllTimeCountHigh)
        tvAllTimeAvg = findViewById(R.id.tvAllTimeAvg)
        tvAllTimeMin = findViewById(R.id.tvAllTimeMin)
        tvAllTimeMax = findViewById(R.id.tvAllTimeMax)
        tvCurWeekCount = findViewById(R.id.tvCurWeekCount)
        tvCurWeekCountLow = findViewById(R.id.tvCurWeekCountLow)
        tvCurWeekCountHigh = findViewById(R.id.tvCurWeekCountHigh)
        tvCurWeekAvg = findViewById(R.id.tvCurWeekAvg)
        tvCurWeekMin = findViewById(R.id.tvCurWeekMin)
        tvCurWeekMax = findViewById(R.id.tvCurWeekMax)
        tvCurMonthCount = findViewById(R.id.tvCountCurMonth)
        tvCurMonthCountLow = findViewById(R.id.tvCurMonthCountLow)
        tvCurMonthCountHigh = findViewById(R.id.tvCurMonthCountHigh)
        tvCurMonthAvg = findViewById(R.id.tvCurMonthAvg)
        tvCurMonthMin = findViewById(R.id.tvCurMonthMin)
        tvCurMonthMax = findViewById(R.id.tvCurMonthMax)
        tvLastWeekCount = findViewById(R.id.tvCountLastWeek)
        tvLastWeekCountLow = findViewById(R.id.tvLastWeekCountLow)
        tvLastWeekCountHigh = findViewById(R.id.tvLastWeekCountHigh)
        tvLastWeekAvg = findViewById(R.id.tvLastWeekAvg)
        tvLastWeekMin = findViewById(R.id.tvLastWeekMin)
        tvLastWeekMax = findViewById(R.id.tvLastWeekMax)
        tvLastMonthCount = findViewById(R.id.tvCountLastMonth)
        tvLastMonthCountLow = findViewById(R.id.tvLastMonthCountLow)
        tvLastMonthCountHigh = findViewById(R.id.tvLastMonthCountHigh)
        tvLastMonthAvg = findViewById(R.id.tvLastMonthAvg)
        tvLastMonthMin = findViewById(R.id.tvLastMonthMin)
        tvLastMonthMax = findViewById(R.id.tvLastMonthMax)
    }

    companion object {
        private const val TAG = "myLog"

        // select query vars
        private const val KEY_COUNT = "COUNT"
        private const val KEY_AVG = "AVG"
        private const val KEY_MIN = "MIN"
        private const val KEY_MAX = "MAX"
        private const val KEY_LOW_SUGAR = true
        private const val KEY_HIGH_SUGAR = false
    }
}